### ISSUE TITLE
<release> versions are not in order [2.10.0 before 2.10.0-rc2]

### ISSUE THREAD

**Issue by Jehan** [hughsie/appstream-glib#234]
Putting 2.10.0 version after 2.10.0-rc* is not working. It seems appstream-util considers 2.10.0 should be considered older than 2.10.0-rc2?

> $ appstream-util validate-relax  desktop/org.gimp.GIMP.appdata.xml
> desktop/org.gimp.GIMP.appdata.xml: FAILED:
> • tag-invalid           : <release> versions are not in order [2.10.0 before 2.10.0-rc2]
> Validation of files failed

**Comment by hughsie** [comment]
Well, that's true isn't it?

    $ rpmdev-vercmp 2.10.0 2.10.0-rc2
    2.10.0 < 2.10.0-rc2

**Comment by ximion** [comment]
@Jehan Try `2.10.0~rc2`, that should be considered smaller than `2.10.0`.

**Comment by Jehan** [comment]
@ximion Thanks I did try a bunch of characters (space, underscore, etc.), but not tilde. It does work for rpmdev-vercmp:

> $ rpmdev-vercmp 2.10.0 2.10.0~rc2
> 2.10.0 > 2.10.0~rc2

But appstream util still fails with the same error:

> $ appstream-util validate-relax desktop/org.gimp.GIMP.appdata.xml
> desktop/org.gimp.GIMP.appdata.xml: FAILED:
> • tag-invalid           : <release> versions are not in order [2.10.0 before 2.10.0~rc2]
> Validation of files failed

@hughsie ? :-)

**Comment by Jehan** [comment]
For the record, I just checked the spec of semantic versioning: https://semver.org/
And it clearly specifies that `2.10.0-rc2 < 2.10.0`.
Cf. point 9. about pre-release version: https://semver.org/#spec-item-9

> A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version.

So aren't actually `rpmdev-vercmp` and appstream-glib wrong about their reverse order? Or are they based on another versioning scheme? (if so, which one?)

**Comment by ximion** [comment]
`rpmdev-vercmp` is correct, appstream itself is as well (uses the same algorithm), but appstream-glib isn't.
The number after the dash is reserved for distribution revisions, e.g. a package revision number or a bundle revision number, and we want to keep compatibility with that.
So, use a tilde instead of a dash, that will be fine. The appstream-glib algorithm should likely be tweaked a bit to support that.

Using AppStream itself:
```
~$ appstreamcli compare-versions 2.10.0 2.10.0-rc2
2.10.0 << 2.10.0-rc2
~$ appstreamcli compare-versions 2.10.0 2.10.0~rc2
2.10.0 >> 2.10.0~rc2
```

**Comment by smcv** [comment]
> For the record, I just checked the spec of semantic versioning

Semver is not *the* spec, it's *a* spec. There is no single canonical specification for comparing version numbers; semver is one possible version comparison algorithm, glibc `strverscmp()` is another, RPM has another, and dpkg is different again (although I think RPM and dpkg are quite similar, and in particular they both give `~` the same special meaning).

If AppStream needs to compare version numbers, then it should define an algorithm or spec to do so. The spec used by dpkg is [in Debian Policy](https://www.debian.org/doc/debian-policy/#version).

**Comment by Jehan** [comment]
> If AppStream needs to compare version numbers, then it should define an algorithm or spec to do so.

Of course. I personally don't care which spec is used. But it would be easier if at least we knew the rules of the spec. :-)
Anyway for now it's alright. No need to just make it more complicated than necessary. Having appstream-glib fixed regarding this tilde separator will likely be enough for us for a few years. :P

**Comment by hughsie** [comment]
Does that look sane? ^^^

**Comment by ximion** [comment]
Makes sense to me :-)
(you could also simply copy the RPM-based version comparison algorithm from libappstream, if you wanted though ;-) )

@smcv I didn't know that `strverscmp` existed, in fact I recently wondered why glibc wouldn't have a version comparison function... Turns out, it exists and I should have looked for that ^^

**Comment by Conan-Kudo** [comment]
@smcv These days, I think RPM and dpkg basically sort the same way, though I like the cleaner implementation in RPM. :)

**Comment by ImUrX** [comment]
i will just replace `-` for `~` in only the appstream in my project, but it would be nice to be compatible with semver qwq
